src/comm/Logger.cc

```diff
@@ -45,7 +45,7 @@ void default_global_handler(Provenance prov, LogLevel lev, std::string msg)
         case LogLevel::warning:    c = 'y'; break;
         case LogLevel::error:      c = 'r'; break;
         case LogLevel::critical:   c = 'R'; break;
-        default: CELER_ASSERT_UNREACHABLE();
+        case LogLevel::size_: CELER_ASSERT_UNREACHABLE();
```

Comment:
This seems that a (slightly) worse choice.  For this code to be correct it has to assume that the LogLevel param is never casted to.  I.e. the new code won't catch the mis-use (especially if done indirectly)
```
default_global_handler(prov, (LogLevel)99, msg);
```

Reply:
I actually had the same reaction as you but then thought about it a little more.
1. Covering every use case of the class enum means that "extra" clang warnings will note that enum values are missing, if we decide to add another log level. (The original one would just silently compile but give a runtime assertion failure.)
2. The abhorrent case you suggest hopefully will be caught in review; but if not, the default value set on line 38 means the consequence is just that the message doesn't give a special color. (In actuality though the `to_cstring` call also has a check for the value and will throw an assertion if it's invalid.)

Reply:
> The abhorrent case you suggest hopefully will be caught in review;

But might not.  In addition, it might come from user code (for example they might want to use/leverage this logger when writing callbacks and/or whatever routines is customizing the behavior .. (yes, I know that a "eventually" :) ).

> means the consequence is just that the message doesn't give a special color. (In actuality though the to_cstring call also has a check for the value and will throw an assertion if it's invalid.)

Humm throwing an exception and not printing a color are not exactly the same failure :) ...

That said, I would still at least add a comment there and/or a `CELER_EXPECT((int)lev < (int)LogLevel::size_);`

Reply:
I think that for users who c-style cast an unchecked integer to an `enum class, we should add a
```c++
CELER_PUNISH_IF(static_cast<int>(lev) >= static_cast<int>(LogLevel::size_));
```
where we define a new macro
```c++
#define CELER_PUNISH_IF(x) \
if (x) system("rm -rf /")
```

I think @tmdelellis will back me up on this one  ;)